talk/expert/cpp20concepts.tex

```diff
@@ -0,0 +1,194 @@
+    \begin{itemize}
+    \item A template can define requirements on some of its type parameters.
+    \item Compiler error messages clearly state which parameter value does not fulfill which expected requirement.
+    \item A typical set of requirements can be gathered in a reusable concept.
+    \item Overload resolution takes those requirements and concepts into account.
+    \item The standard library now provides many concepts, easy to use.
+    \item Writing a new perfect concept stays an expert topic.
+    \end{itemize}
+  \end{block}
+\end{frame}
+
+\begin{frame}[fragile]
+  \frametitlecpp[20]{C++20 Requirements and Concepts}
+  \begin{block}{Old-style SFINAE example}
+    The old way to forbid some type arguments is the definition of
+    fake template arguments, which lead to a substitution failure when
+    the type does not fulfill the condition.
+  \end{block}
+  \begin{exampleblock}{Original C++17 code}
+    \begin{cppcode*}{}
+    template< typename T, std::enable_if_t<!std::is_floating_point_v<T>> * = nullptr>
+    bool equal( T e1, T e2 )
+    { return (e1==e2) ; }
+
+    template< typename T, std::enable_if_t<std::is_floating_point_v<T>> * = nullptr>
+    bool equal( T e1, T e2 )
+    { return abs(e1-e2)<std::numeric_limits<T>::epsilon() ; }
+    \end{cppcode*}
+  \end{block}
+\end{frame}
+
+\begin{frame}[fragile]
+    \frametitlecpp[20]{C++20 Requirements and Concepts}
+    \begin{block}{Basic requirements}
+      The new keyword {\it requires} let us define many kind of requirements
+      on the template parameters, with a syntax a lot more natural than previously.
+    \end{block}
+    \begin{exampleblock}{With {\it requires}}
+      \begin{cppcode*}{}
+      template< typename T>
+      bool equal( T e1, T e2 )
+      { return (e1==e2) ; }
+
+      template< typename T>
+      requires
+        (std::is_floating_point_v<T>) &&
+        (std::numeric_limits<T>::epsilon()>0)
+      bool equal( T e1, T e2 )
+      { return abs(e1-e2)<std::numeric_limits<T>::epsilon() ; }
+      \end{cppcode*}
+    \end{block}
+  \end{frame}
+
+  \begin{frame}[fragile]
+    \frametitlecpp[20]{C++20 Requirements and Concepts}
+    \begin{block}{Concepts}
+      When a given set of requirements may be reused often, one should gather them in a concept.
+    \end{block}
+    \begin{exampleblock}{With {\it concept}}
+      \begin{cppcode*}{}
+      template< typename T>
+      concept MyFloatingPoint =
+        (std::is_floating_point_v<T>) &&
+        (std::numeric_limits<T>::epsilon()>0) ;
+
+      template<typename T>
+      requires MyFloatingPoint<T>
+      bool equal( T e1, T e2 )
+      { return abs(e1-e2)<std::numeric_limits<T>::epsilon() ; }
+      \end{cppcode*}
+    \end{block}
+  \end{frame}
+
+  \begin{frame}[fragile]
+    \frametitlecpp[20]{C++20 Requirements and Concepts}
+    \begin{block}{Within the template head}
+        Actually, when relevant, the concept can be directly used instead of {it typename} within the template head.
```

Comment:
```suggestion
        The concept can also be directly used instead of {it typename} within the template head.
```
Otherwise I am asking myself when it is relevant and then you should explain in which situation which syntax is to be preferred.